services/src/main/java/org/keycloak/vault/DefaultVaultRawSecret.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.vault;
+
+import java.util.Optional;
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * Default raw secret implementation for {@code byte[]}.
+ * @author hmlnarik
+ */
+public class DefaultVaultRawSecret implements VaultRawSecret {
+
+    private final Optional<byte[]> rawSecret;
```

Comment:
In general, I don't like `Optional`s as field. We've had a lengthy discussion about them in Infinispan. A very quick summary is the following:
- `Optional`s don't implement `Serializable`, their intention is to be used as a return type, not as a field.
- `Optional`s used as fields are extremely likely (unless you're using non-standard JVM parameters) to be JIT'ted out. The C2 compiler eliminates even the object creation.
- When you use `Optional` as a field, it's very likely that it will cost you at least an object allocation. The `Optional` object will stay there until it's `reachable` from GC perspective. Of course, if we assume that `DefaultVaultRawSecret` is a short-lived object, that doesn't apply here.

Of course, I'm not insisting to change it here (I've taken part in a flame war around `Optional`s once and I'm not doing this again). I can give you some links if you wish but the `Optional` was often turns into `tabs vs spaces`-type of war.

But as a final thought - you do check if the `Optional` is present and then if the `buffer` is not null, which would get even simpler is the `buffer` was just a nullable `byte[]`.

Reply:
These are good points. I've updated the default implementation accordingly.